### PR TITLE
Re-apply remaining changes from limit_dataflow_recursion branch

### PR DESCRIPTION
**Pull request by hkaiser**
- fix access to recursion count for non-HPX-threads
- re-add test verifying dataflow recursion does not cause stack
  overflows
- add test for exception propagation from continuations

### CODE REVIEW

**Comment by sithhell** [comment]
I am confused, the unit and regression tests seem to pass OK now, we do we need to limit recursion again?

File: hpx/lcos/detail/future_data.hpp [hunk 1]
```diff
@@ -219,37 +240,98 @@ namespace detail
...
-                // handle all threads waiting for the block to become full
-                cond_.notify_all(std::move(l), ec);
-            }
+            // handle all threads waiting for the future to become ready
+            cond_.notify_all(std::move(l), ec);
 
             // invoke the callback (continuation) function
             if (on_completed)
-                on_completed();
+                handle_on_completed(std::move(on_completed));
```
Comment (sithhell) [inline hpx/lcos/detail/future_data.hpp]: This is not good. The user provided completion handler is called while a lock is held.

Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: No, `cond_.notify_all()` 'consumes' the lock and leaves it unlocked. I'll add a comment to this code.

Comment (sithhell) [inline hpx/lcos/detail/future_data.hpp]: Yes please, this is a very odd looking code construct. But seems that it works as intended.

**Comment by sithhell** [comment]
Looks good to merge after the comment has been added.

**Comment by hkaiser** [comment]
> I am confused, the unit and regression tests seem to pass OK now, we do we need to limit recursion again?

This patch contains a test which would cause stack overflows without it.

**Comment by hkaiser** [comment]
All comments have been addressed

**Comment by sithhell** [comment]
Thanks, looking good now!

**Pull request merged**